[TypeDeclaration] ReturnTypeDeclarationRector explodes with anonymous class instances being returned

## Current Behaviour 

Rector: `Rector\TypeDeclaration\Rector\FunctionLike\ReturnTypeDeclarationRector`

I'm not sure if the error is on rector itself or on phpstan, but when running the code below, I'm getting the following error:

> [ERROR] Could not process "file.php" file, due to: "Internal error.".                                                                                             

The callstack is something like the following:
```
PHPStan\ShouldNotHappenException: Internal error.

/rector/vendor/phpstan/phpstan/src/Broker/Broker.php:294
/rector/vendor/phpstan/phpstan/src/Analyser/Scope.php:1093
/rector/vendor/phpstan/phpstan/src/Analyser/Scope.php:364
/rector/packages/NodeTypeResolver/src/NodeTypeResolver.php:300
/rector/packages/NodeTypeResolver/src/NodeTypeResolver.php:258
/rector/packages/NodeTypeResolver/src/NodeTypeResolver.php:308
/rector/src/PhpParser/Node/Manipulator/FunctionLikeManipulator.php:65
/rector/packages/TypeDeclaration/src/Rector/FunctionLike/ReturnTypeDeclarationRector.php:134
/rector/packages/TypeDeclaration/src/Rector/FunctionLike/ReturnTypeDeclarationRector.php:85
```                             

I'm not quite sure how to fix it, I've tried to take a look at phpstan but it seems too complex for me. So I'm asking for directions here fist.

## Minimal PHP Code Causing Issue

```php
public function getObject()
{
    return new class() {};
}
```

## Expected Behaviour

```diff
-public function getObject()
+public function getObject(): object
 {
     return new class() {};
 }
```

--- a/packages/NodeTypeResolver/src/NodeTypeResolver.php
+++ b/packages/NodeTypeResolver/src/NodeTypeResolver.php
@@ -35,6 +35,7 @@ use PHPStan\Type\IntersectionType;
 use PHPStan\Type\MixedType;
 use PHPStan\Type\NullType;
 use PHPStan\Type\ObjectType;
+use PHPStan\Type\ObjectWithoutClassType;
 use PHPStan\Type\StringType;
 use PHPStan\Type\Type;
 use PHPStan\Type\UnionType;
@@ -297,6 +298,12 @@ final class NodeTypeResolver
             return null;
         }
 
+        if ($node instanceof New_) {
+            if ($this->isAnonymousClass($node->class)) {
+                return new ObjectWithoutClassType();
+            }
+        }
+
         return $nodeScope->getType($node);
     }
 
@@ -410,8 +417,7 @@ final class NodeTypeResolver
 
         // skip anonymous classes, ref https://github.com/rectorphp/rector/issues/1574
         if ($node instanceof New_) {
-            $className = $this->nameResolver->resolve($node->class);
-            if ($className === null || Strings::contains($className, 'AnonymousClass')) {
+            if ($this->isAnonymousClass($node->class)) {
                 return [];
             }
         }
@@ -555,5 +561,16 @@ final class NodeTypeResolver
         }
 
         return null;
+    }
+
+    private function isAnonymousClass(Node $node): bool
+    {
+        if (! $node instanceof Class_) {
+            return false;
+        }
+
+        $className = $this->nameResolver->resolve($node);
+
+        return $className === null || Strings::contains($className, 'AnonymousClass');
     }
 }

--- a/packages/NodeTypeResolver/src/StaticTypeToStringResolver.php
+++ b/packages/NodeTypeResolver/src/StaticTypeToStringResolver.php
@@ -10,6 +10,7 @@ use PHPStan\Type\FloatType;
 use PHPStan\Type\IntegerType;
 use PHPStan\Type\NullType;
 use PHPStan\Type\ObjectType;
+use PHPStan\Type\ObjectWithoutClassType;
 use PHPStan\Type\StringType;
 use PHPStan\Type\Type;
 use PHPStan\Type\UnionType;
@@ -26,6 +27,7 @@ final class StaticTypeToStringResolver
     {
         $resolvers = [
             IntegerType::class => ['int'],
+            ObjectWithoutClassType::class => ['object'],
             ClosureType::class => ['callable'],
             CallableType::class => ['callable'],
             FloatType::class => ['float'],

--- a/packages/TypeDeclaration/tests/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/a_new_class.php.inc
+++ b/packages/TypeDeclaration/tests/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/a_new_class.php.inc
@@ -9,3 +9,19 @@ class ANewClass
         return new class() {};
     }
 }
+
+?>
+-----
+<?php
+
+namespace Rector\TypeDeclaration\Tests\Rector\ClassMethod\ReturnTypeDeclarationRector\Fixture;
+
+class ANewClass
+{
+    public function getObject(): object
+    {
+        return new class() {};
+    }
+}
+
+?>

--- a/packages/TypeDeclaration/tests/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/a_new_class.php.inc
+++ b/packages/TypeDeclaration/tests/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/a_new_class.php.inc
@@ -1,0 +1,11 @@
+<?php
+
+namespace Rector\TypeDeclaration\Tests\Rector\ClassMethod\ReturnTypeDeclarationRector\Fixture;
+
+class ANewClass
+{
+    public function getObject()
+    {
+        return new class() {};
+    }
+}

--- a/packages/TypeDeclaration/tests/Rector/FunctionLike/ReturnTypeDeclarationRector/ReturnTypeDeclarationRectorTest.php
+++ b/packages/TypeDeclaration/tests/Rector/FunctionLike/ReturnTypeDeclarationRector/ReturnTypeDeclarationRectorTest.php
@@ -53,6 +53,8 @@ final class ReturnTypeDeclarationRectorTest extends AbstractRectorTestCase
             __DIR__ . '/Fixture/dunglas/BazTrait.php.inc',
             __DIR__ . '/Fixture/dunglas/Child.php.inc',
             __DIR__ . '/Fixture/dunglas/nullable_types.php.inc',
+            // anonymous class
+            __DIR__ . '/Fixture/a_new_class.php.inc',
         ];
 
         $this->doTestFiles($integrationFiles);

